### PR TITLE
Add MACsec Access modules:

### PR DESCRIPTION
**Pull request by therealdoug**
- ADD: `modules/terraform-aci-macsec-interfaces-policy`
- ADD: `modules/terraform-aci-macsec-keychain-policies`
- ADD: `modules/terraform-aci-macsec-parameters-policy`
- UPDATE: aci_access_policies.yaml
- UPDATE: terraform-aci-access-leaf-interface-policy-group
- UPDATE: defaults.yaml
- UPDATE: modules.yaml

### CODE REVIEW

**Comment by therealdoug** [comment]
@juchowan pre-commits are passing but failing tests.  issue with the apic sandbox?

**Comment by danischm** [comment]
Indeed.. I have restarted the workflow and it was successful now.

File: aci_access_policies.tf [hunk 1]
```diff
@@ -301,6 +301,62 @@ module "aci_link_level_policy" {
...
+}
+
+locals {
+  macsec_keychain_policies = flatten([
+    for mkc in try(local.access_policies.interface_policies.macsec_keychain_policies, []) : {
+      name        = "${mkc.name}${local.defaults.apic.access_policies.interface_policies.macsec_keychain_policies.name_suffix}"
+      description = try(mkc.description, "")
+      key_policies = [for kp in try(mkc.key_policies, []) : {
+        name         = kp.name
+        keyName      = kp.keyName
```
Comment (juchowan) [inline aci_access_policies.tf:325]: all our variables should be in the format of x_y, so key_name in this case

File: aci_access_policies.tf [hunk 2]
```diff
@@ -301,6 +301,62 @@ module "aci_link_level_policy" {
...
+        preSharedKey = kp.preSharedKey
```
Comment (juchowan) [inline aci_access_policies.tf:326]: same as above

File: aci_access_policies.tf [hunk 3]
```diff
@@ -301,6 +301,62 @@ module "aci_link_level_policy" {
...
+        description  = try(kp.description, "")
+        startTime    = try(kp.startTime, local.defaults.apic.access_policies.interface_policies.macsec_keychain_policies.startTime)
```
Comment (juchowan) [inline aci_access_policies.tf:328]: same as above

File: aci_access_policies.tf [hunk 4]
```diff
@@ -301,6 +301,62 @@ module "aci_link_level_policy" {
...
+        endTime      = try(kp.endTime, local.defaults.apic.access_policies.interface_policies.macsec_keychain_policies.endTime)
```
Comment (juchowan) [inline aci_access_policies.tf:329]: same as above

File: aci_access_policies.tf [hunk 5]
```diff
@@ -392,6 +448,7 @@ module "aci_access_leaf_interface_policy_group" {
   cdp_policy                 = try("${each.value.cdp_policy}${local.defaults.apic.access_policies.interface_policies.cdp_policies.name_suffix}", "")
   lldp_policy                = try("${each.value.lldp_policy}${local.defaults.apic.access_policies.interface_policies.lldp_policies.name_suffix}", "")
   spanning_tree_policy       = try("${each.value.spanning_tree_policy}${local.defaults.apic.access_policies.interface_policies.spanning_tree_policies.name_suffix}", "")
+  macsec_policy              = try(each.value.macsec_policy, "")
```
Comment (juchowan) [inline aci_access_policies.tf:451]: should include the name suffix for macsec policy

File: modules/terraform-aci-access-leaf-interface-policy-group/main.tf [hunk 6]
```diff
@@ -36,6 +36,15 @@ resource "aci_rest_managed" "infraRsLldpIfPol" {
   }
 }
 
+resource "aci_rest_managed" "infraRsMacsecIfPol" {
+  count      = var.type != "breakout" ? 1 : 0
```
Comment (juchowan) [inline modules/terraform-aci-access-leaf-interface-policy-group/main.tf:40]: i think this should check macsec policy, not breakout

File: modules/terraform-aci-macsec-interfaces-policy/examples/complete/.terraform-docs.yml
Comment (juchowan) [inline modules/terraform-aci-macsec-interfaces-policy/examples/complete/.terraform-docs.yml:1]: documentation for example should follow other format than documentation of examples, please check on other existing modules

File: modules/terraform-aci-macsec-interfaces-policy/main.tf [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  class_name = "macsecIfPol"
+
+  content = {
+    name    = var.name
+    adminSt = var.admin_state ? "enabled" : "disabled"
+  }
+}
+
+resource "aci_rest_managed" "macsecRsToKeyChainPol" {
+  dn         = "uni/infra/macsecifp-${var.name}/rsToKeyChainPol"
```
Comment (juchowan) [inline modules/terraform-aci-macsec-interfaces-policy/main.tf:12]: change to 
dn = "${aci_rest_managed.macsecIfPol.dn}/rsToKeyChainPol
and delete the dependencies

File: modules/terraform-aci-macsec-interfaces-policy/main.tf [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  content = {
+    tDn = "uni/infra/macsecpcont/keychainp-${var.macsec_keychain_policy}"
+  }
+
+  depends_on = [
+    aci_rest_managed.macsecIfPol
+  ]
+}
+
+resource "aci_rest_managed" "macsecRsToParamPol" {
```
Comment (juchowan) [inline modules/terraform-aci-macsec-interfaces-policy/main.tf:24]: same for this resource, change DN to refer to macsecIfPol resource

File: modules/terraform-aci-macsec-interfaces-policy/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,9 @@
+output "id" {
```
Comment (juchowan) [inline modules/terraform-aci-macsec-interfaces-policy/outputs.tf:1]: output "dn"

File: modules/terraform-aci-macsec-keychain-policies/examples/complete/.terraform-docs.yml
Comment (juchowan) [inline modules/terraform-aci-macsec-keychain-policies/examples/complete/.terraform-docs.yml:1]: example documentation follows different format

File: modules/terraform-aci-macsec-keychain-policies/main.tf [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  class_name = "macsecKeyChainPol"
+  content = {
+    name  = var.name
+    descr = var.description
+  }
+}
+
+resource "aci_rest_managed" "macsecKeyPol" {
+  for_each   = { for kp in var.key_policies : kp.keyName => kp }
+  dn         = "uni/infra/macsecpcont/keychainp-${var.name}/keyp-${each.value.keyName}"
```
Comment (juchowan) [inline modules/terraform-aci-macsec-keychain-policies/main.tf:13]: change dn to reflect macsecKeyChainPol resource

File: modules/terraform-aci-macsec-keychain-policies/outputs.tf
Comment (juchowan) [inline modules/terraform-aci-macsec-keychain-policies/outputs.tf:1]: add dn output

File: modules/terraform-aci-macsec-keychain-policies/variables.tf [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+  validation {
+    condition     = can(regex("^[a-zA-Z0-9\\!#$%()*,-./:;@ _{|}~?&+]{0,128}$", var.description))
+    error_message = "Allowed characters: `a`-`z`, `A`-`Z`, `0`-`9`, `\\`, `!`, `#`, `$`, `%`, `(`, `)`, `*`, `,`, `-`, `.`, `/`, `:`, `;`, `@`, ` `, `_`, `{`, `|`, }`, `~`, `?`, `&`, `+`. Maximum characters: 128."
+  }
+}
+
+variable "key_policies" {
+  description = "Key Polices for Key Chain"
+  type        = list(map(string))
```
Comment (juchowan) [inline modules/terraform-aci-macsec-keychain-policies/variables.tf:24]: shouldn't this list all options available under key_policies with its types?

File: modules/terraform-aci-macsec-parameters-policy/examples/complete/.terraform-docs.yml
Comment (juchowan) [inline modules/terraform-aci-macsec-parameters-policy/examples/complete/.terraform-docs.yml:1]: different format

File: modules/terraform-aci-macsec-parameters-policy/examples/complete/main.tf
Comment (juchowan) [inline modules/terraform-aci-macsec-parameters-policy/examples/complete/main.tf:1]: example doesn't reflect what variables are expected, confOffset -> confidentiality_offset etc

**Review by juchowan — CHANGES_REQUESTED** [review]

File: modules/terraform-aci-access-leaf-interface-policy-group/main.tf (hunk 6 above)
Comment (therealdoug) [reply modules/terraform-aci-access-leaf-interface-policy-group/main.tf:40]: Documentation isn't clear, is macsec supported breakout dynamic breakout ports?

**Comment by therealdoug** [comment]
Thanks for the feedback, I believe comments are resolved!

File: modules/terraform-aci-macsec-keychain-policies/main.tf [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+resource "aci_rest_managed" "macsecKeyPol" {
+  for_each   = { for kp in var.key_policies : kp.key_name => kp }
+  dn         = "${aci_rest_managed.macsecKeyChainPol.dn}/keyp-${each.value.key_name}"
+  class_name = "macsecKeyPol"
+  content = {
+    name         = each.value.name
+    keyName      = each.value.key_name
+    descr        = each.value.description
+    preSharedKey = each.value.pre_shared_key
```
Comment (juchowan) [inline modules/terraform-aci-macsec-keychain-policies/main.tf:19]: this should be sensitive value

File: aci_access_policies.tf [hunk 13]
```diff
@@ -301,6 +301,62 @@ module "aci_link_level_policy" {
...
+  for_each     = { for mkc in try(local.macsec_keychain_policies, []) : mkc.name => mkc if local.modules.aci_macsec_keychain_policies && var.manage_access_policies }
+  name         = "${each.value.name}${local.defaults.apic.access_policies.interface_policies.macsec_keychain_policies.name_suffix}"
+  description  = each.value.description
+  key_policies = each.value.key_policies
+}
+
+module "aci_macsec_interfaces_policy" {
+  source = "./modules/terraform-aci-macsec-interfaces-policy"
+
+  for_each                 = { for mip in try(local.access_policies.interface_policies.macsec_interfaces_policy, []) : mip.name => mip if local.modules.aci_macsec_interfaces_policy && var.manage_access_policies }
```
Comment (juchowan) [inline aci_access_policies.tf:348]: should be plural: local.access_policies.interface_policies.macsec_interfaces_policies

File: aci_access_policies.tf [hunk 14]
```diff
@@ -392,6 +448,7 @@ module "aci_access_leaf_interface_policy_group" {
   cdp_policy                 = try("${each.value.cdp_policy}${local.defaults.apic.access_policies.interface_policies.cdp_policies.name_suffix}", "")
   lldp_policy                = try("${each.value.lldp_policy}${local.defaults.apic.access_policies.interface_policies.lldp_policies.name_suffix}", "")
   spanning_tree_policy       = try("${each.value.spanning_tree_policy}${local.defaults.apic.access_policies.interface_policies.spanning_tree_policies.name_suffix}", "")
+  macsec_policy              = try("${each.value.macsec_policy}${local.defaults.apic.access_policies.interface_policies.macsec_interfaces_policy.name_suffix}", "")
```
Comment (juchowan) [inline aci_access_policies.tf:451]: please change to macsec_interface_policy

**Review by juchowan — CHANGES_REQUESTED** [review]

**Review by therealdoug — COMMENTED** [review]
Updated and completed.

**Comment by therealdoug** [comment]
Closing PR.  Will submit new PR with fabric macsec policies included.

**Comment by juchowan** [comment]
can you reopen this? we can still use it in the future

**Pull request closed without merging**